mlfinlab/portfolio_optimization/risk_estimators.py

```diff
@@ -0,0 +1,259 @@
+        pdf = tn_relation * ((eigen_max - eigen_space) * (eigen_space - eigen_min)) ** (1 / 2) / \
+                             (2 * np.pi * var * eigen_space)
+        pdf = pd.Series(pdf, index=eigen_space)
+
+        return pdf
+
+    def _pdf_fit(self, var, eigen_observations, tn_relation, kde_bwidth, num_points=1000):
+        """
+        Calculates the fit (Sum of Squared estimate of Errors) of the empirical pdf
+        (kernel density estimation) to the theoretical pdf (Marcenko-Pastur distribution).
+
+        SSE is calculated for num_points, equally spread between minimum and maximum
+        expected theoretical eigenvalues.
+
+        :param var: (float) Variance of the M-P distribution (for the theoretical pdf)
+        :param eigen_observations: (np.array) Observed empirical eigenvalues (for the empirical pdf)
+        :param tn_relation: (float) Relation of sample length T to the number of variables N (for the theoretical pdf)
+        :param kde_bwidth: (float) The bandwidth of the kernel (for the empirical pdf)
+        :param num_points: (int) Number of points to estimate pdf  (for the empirical pdf)
+        :return: (float) SSE between empirical pdf and theoretical pdf
+        """
+
+        # Calculating theoretical and empirical pdf
+        theoretical_pdf = self._mp_pdf(var, tn_relation, num_points)
+        empirical_pdf = self._fit_kde(eigen_observations, kde_bwidth, eval_points=theoretical_pdf.index.values)
+
+        # Fit calculation
+        sse = np.sum((empirical_pdf - theoretical_pdf) ** 2)
+
+        return sse
+
+    def _find_max_eval(self, eigen_observations, tn_relation, kde_bwidth):
+        """
+        Searching for maximum random eigenvalue by fitting Marcenko-Pastur distribution
+        to the empirical one - obtained through kernel density estimation.
+
+        :param eigen_observations: (np.array) Observed empirical eigenvalues (for the empirical pdf)
+        :param tn_relation: (float) Relation of sample length T to the number of variables N (for the theoretical pdf)
+        :param kde_bwidth: (float) The bandwidth of the kernel (for the empirical pdf)
+        :return: (float, float) Maximum random eigenvalue, optimal variation of the Marcenko-Pastur distribution
+        """
+
+        # Searching for the variation of Marcenko-Pastur distribution for the best fit with the empirical distribution
+        optimization = minimize(self._pdf_fit, x0=np.array(0.5), args=(eigen_observations, tn_relation, kde_bwidth),
+                                bounds=((1e-5, 1 - 1e-5),))
+
+        # The optimal solution found
+        var = optimization['x'][0]
+
+        # Eigenvalue calculated as the maximum expected eigenvalue based on the input
+        maximum_eigen = var * (1 + (1 / tn_relation) ** (1 / 2)) ** 2
+
+        return maximum_eigen, var
+
+    @staticmethod
+    def corr_to_cov(corr, std):
+        """
+        Recovers the covariance matrix from the de-noise correlation matrix.
```

Comment:
Also, is this function particular to denoising covariance method? Can this not be used to retrieve any covariance matrix from a correlation matrix?

If yes, then you can change the docstring to remove the reference to de-noise matrix

Reply:
You're right. I'll change the the docstring also since it's a public function now.

Reply:
Done